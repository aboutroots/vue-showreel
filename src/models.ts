/* eslint-disable camelcase */

export interface RootStore {
  modules: {
    auth: Record<string, unknown>;
    ui: Record<string, unknown>;
    users: Record<string, unknown>;
  };
}

export interface UserDTO {
  id: number;
  email: string;
  first_name: string;
  last_name: string;
  avatar: string;
}
