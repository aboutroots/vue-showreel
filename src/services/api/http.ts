import axios from 'axios';

const apiClient = axios.create({
  baseURL: 'https://reqres.in/api/',
  headers: {},
  maxRedirects: 0,
});

export default apiClient;
