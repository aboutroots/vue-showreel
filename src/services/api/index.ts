import usersApi from './resources/users';
import authApi from './resources/auth';

export default {
  users: usersApi,
  auth: authApi,
};
