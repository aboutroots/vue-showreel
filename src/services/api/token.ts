import store from 'store';

/**
 * Utility facade for storing authorization token in localStorage
 * on different devices via storejs package
 */
class TokenManager {
  token: string | null;

  constructor() {
    this.token = store.get('token', null);
  }

  set(token: string): void {
    this.token = token;
    store.set('token', token);
  }

  get(): string | null {
    return this.token;
  }

  delete(): void {
    store.remove('token');
    this.token = null;
  }
}

export default new TokenManager();
