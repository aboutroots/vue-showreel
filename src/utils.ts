export const sleep = async (ms: number): Promise<void> => {
  /**
   * Sleep for given number of miliseconds
   */
  // eslint-disable-next-line no-promise-executor-return
  await new Promise((r) => setTimeout(r, ms));
};
