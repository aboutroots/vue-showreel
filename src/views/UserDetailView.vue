<template>
  <section class="container">
    <h1>User Detail View</h1>
    <Loader v-if="processing" />
    <div v-else class="user-detail fadeInBasic">
      <img :src="user.avatar" class="avatar" alt="User avatar" />
      <h3 class="title">
        <span class="last">{{ user.last_name }}</span
        >, {{ user.first_name }}
      </h3>
      <p class="subtitle">{{ user.email }}</p>
      <div class="actions">
        <button class="btn">+ FOLLOW</button>
        <button class="btn"><ios-mail-icon class="btn-icon" w="20" h="20" />SEND MESSAGE</button>
      </div>
      <img
        src="@/assets/img/collab.svg"
        class="bg-img"
        alt="Background image with people collaborating"
      />
    </div>
  </section>
</template>

<script lang="ts">
import Loader from '@/components/SimpleLoader.vue';

import Vue from 'vue';
import { UserDTO } from '@/models';
import api from '../services/api';

interface State {
  processing: boolean;
  user: UserDTO;
}

export default Vue.extend({
  name: 'UserDetailView',
  components: {
    Loader,
  },
  data() {
    return {
      processing: false,
      user: {},
    } as State;
  },
  async mounted(): Promise<void> {
    await this.loadUser();
  },
  methods: {
    async loadUser(): Promise<void> {
      this.processing = true;
      try {
        this.user = await api.users.getUser(this.$route.params.id);
      } catch {
        this.$toast.error('Some error occured!');
      }
      this.processing = false;
    },
  },
});
</script>

<style lang="scss" scoped>
@import '@/assets/styles/theme.scss';
.user-detail {
  margin-top: 5vh;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.avatar {
  border-radius: 50%;
}
.title {
  font-size: 2rem;
  color: $gray;
  .last {
    text-transform: uppercase;
    color: $light;
  }
}
.subtitle {
  text-transform: uppercase;
  color: $gray;
  font-weight: bold;
}
.actions {
  padding: 2rem;
  border-top: 1px solid $gray;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  .btn {
    margin: 0.5rem 2rem;
    .btn-icon {
      margin-right: 0.5rem;
    }
  }
}
.bg-img {
  max-height: 150px;
  margin-top: 2rem;
}
</style>
