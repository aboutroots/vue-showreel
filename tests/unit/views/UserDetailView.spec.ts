import flushPromises from 'flush-promises';
import UserDetailView from '@/views/UserDetailView.vue';
import { UserDTO } from '@/models';
import api from '@/services/api';
import { shallowMount } from '../utils';

jest.mock('@/services/api');
const defaultUser: UserDTO = {
  id: 1,
  email: 'test@test.com',
  first_name: 'John',
  last_name: 'Test',
  avatar: 'https://img.com/1.png',
};

describe('UserDetailView tests suite', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    api.users.getUser.mockResolvedValue(defaultUser);
  });
  test('renders the component successfully', async () => {
    const wrapper = shallowMount(UserDetailView);
    await flushPromises();

    const container = wrapper.find('.UserDetailView');

    expect(container.exists()).toBeTruthy();
  });

  test('matches snapshot', async () => {
    const wrapper = shallowMount(UserDetailView);
    await flushPromises();

    expect(wrapper.element).toMatchSnapshot();
  });
});
