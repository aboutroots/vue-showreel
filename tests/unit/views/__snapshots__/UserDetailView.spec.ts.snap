// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetailView tests suite matches snapshot 1`] = `
<section
  class="UserDetailView container"
>
  <h1>
    User Detail View
  </h1>
   
  <div
    class="user-detail fadeInBasic"
  >
    <img
      alt="User avatar"
      class="avatar shadow"
      src="/img/1.png"
    />
     
    <h3
      class="title"
    >
      <span
        class="last"
      >
        Test
      </span>
      , John
    
    </h3>
     
    <p
      class="subtitle"
    >
      johntest@test.com
    </p>
     
    <div
      class="actions"
    >
      <button
        class="btn"
      >
        + Follow
      </button>
       
      <button
        class="btn"
      >
        <ios-mail-icon-stub
          animate=""
          class="btn-icon"
          h="20"
          rootclass=""
          title=""
          w="20"
        />
        Send message
      </button>
    </div>
     
    <img
      alt="Background image with people collaborating"
      class="bg-img"
      src="@/assets/img/collab.svg"
    />
  </div>
</section>
`;
